src/Compilers/CSharp/Portable/BoundTree/BoundDecisionDag.cs

```diff
@@ -196,6 +194,37 @@ BoundDecisionDagNode makeReplacement(BoundDecisionDagNode dag, Func<BoundDecisio
             }
         }
 
+        /// <summary>
+        /// To facilitate subsumption-checking for list-patterns, some nodes may be inserted into the dag during construction.
+        /// We will remove such nodes immediately after as these are not important in lowering or any other analysis.
+        /// </summary>
+        public BoundDecisionDag RemoveCompilerGeneratedNodes()
+        {
+            return Rewrite(static (node, replacement) =>
+            {
+                if (node.WasCompilerGenerated)
+                {
+                    switch (node)
+                    {
+                        case BoundEvaluationDecisionDagNode evalNode:
+                            Debug.Assert(evalNode.Evaluation is BoundDagAssignmentEvaluation);
```

Comment:
>BoundDagAssignmentEvaluation [](http://example.com/codeflow?start=64&length=28)

Is there ever a reason to keep ```BoundDagAssignmentEvaluation``` nodes in the DAG till lowering?

Reply:
At first I thought this could be used to avoid calling the the next indexer where it's the same as the one we just invoked, but that doesn't work out-of-the-box because of how nodes are ordered and there seem to be not much gain in trying to do so.

Reply:
>>Is there ever a reason to keep ```BoundDagAssignmentEvaluation``` nodes in the DAG till lowering?

> At first I thought this could be used to avoid calling the the next indexer where it's the same as the one we just invoked, but that doesn't work out-of-the-box because of how nodes are ordered and there seem to be not much gain in trying to do so.

Then I suggest the node should be eliminated regardless of its WasCompilerGenerated status.

Reply:
Now I see what you were asking. WasCompilerGenerated is always set for BoundDagAssignmentEvaluation and propagated to the containing node. So yes, it always gets eliminated.

Reply:
> WasCompilerGenerated is always set for BoundDagAssignmentEvaluation and propagated to the containing node. So yes, it always gets eliminated.

To make sure we are on the same page. I think we shouldn't rely on ```WasCompilerGenerated``` to eliminate this node. Lowering cannot handle it in any form.